lib/fastly_nsq/listener.rb

```diff
@@ -1,59 +1,115 @@
-      @processor    = processor
       @consumer     = consumer || FastlyNsq::Consumer.new(topic: topic, channel: channel)
+      @done         = false
       @logger       = options.fetch :logger, FastlyNsq.logger
+      @manager      = options[:manager] || FastlyNsq::Manager.new
       @preprocessor = options[:preprocessor]
+      @processor    = processor
+      @thread       = nil
+      @topic        = topic
     end
 
-    def go(run_once: false)
-      exit_on 'INT'
-      exit_on 'TERM'
+    def identity
+      {
+        consumer:     @consumer,
+        logger:       @logger,
+        manager:      @manager,
+        preprocessor: @preprocessor,
+        processor:    @processor,
+        topic:        @topic,
+      }
+    end
+
+    def clean_dup
+      reset
+      dup
+    end
 
-      loop do
+    def start
+      @logger.info { "> Listener Started: topic #{@topic}" }
+      @thread ||= safe_thread('listener', &method(:go))
+    end
+
+    def go(run_once: false)
+      until @done
         next_message do |message|
           log message
           preprocess message
-          processor.process message
+          @processor.process message
         end
 
-        break if run_once
+        terminate if run_once
       end
 
-      consumer.terminate
+      @manager.listener_stopped(self)
+    rescue FastlyNsq::Shutdown
+      @manager.listener_stopped(self)
+    rescue Exception => ex # rubocop:disable Lint/RescueException
+      @logger.error ex.inspect
+      @manager.listener_killed(self)
+    ensure
+      cleanup
     end
 
-    private
+    def status
+      @thread.status if @thread
+    end
+
+    def terminate
+      @done = true
+      return unless @thread
+      @logger.info "< Listener TERM: topic #{@topic}"
+    end
+
+    def kill
+      @done = true
+      return unless @thread
+      @logger.info "< Listener KILL: topic #{@topic}"
+      @thread.raise FastlyNsq::Shutdown
+    end
 
-    attr_reader :topic, :consumer, :preprocessor, :processor, :logger
+    private
 
     def log(message)
-      logger.info "[NSQ] Message Received: #{message}" if logger
+      @logger.info "[NSQ] Message Received: #{message}" if @logger
+    end
+
+    def cleanup
+      @consumer.terminate
+      @logger.info '< Consumer terminated'
     end
 
     def next_message
-      message = consumer.pop # TODO: consumer.pop do |message|
+      message = @consumer.pop # TODO: consumer.pop do |message|
       result  = yield FastlyNsq::Message.new(message.body)
       message.finish if result
     end
 
     def preprocess(message)
-      preprocessor.call(message) if preprocessor
+      @preprocessor.call(message) if @preprocessor
     end
 
-    def exit_on(signal)
-      Signal.trap(signal) do
-        consumer.terminate
-        exit
-      end
+    def reset
+      @done = false
+      @thread = nil
+      self
     end
   end
 end
+
+class FastlyNsq::Shutdown < StandardError; end
+
+require 'fastly_nsq/listener/config'
```

Comment:
If you change `class Listener::Config` to `class Listener\nclassConfig` it should work as expected. (iirc, `class Foo::Bar` assumes/requires that Foo has already been defined) Alternatively, you could move the `Listener::Config` inline in this file: both classes are reasonably small to make that manageable, imo.